### DISCUSSION TITLE
Combine multiple selectors together

### DISCUSSION THREAD

**Discussion by dudulasry** [TanStack/query#2954] [Ideas]
Hey everyone,

I'm using react-query for a while now, and I was wondering whether the library allows chaining multiple selectors together. This turned out to be a real need in a project I work on.

If this isn't supported, maybe we can allow it by exposing a utility function that combines selectors:
```
const combineSelectors = (selectors) => (data) =>
  selectors.reduce((input, selector) => selector(input), data);
```

And then we can use it in the code like this:
```
const selector = combineSelectors([selector1, selector2]);
```

Another idea is to combine selectors behind the scenes. We can check if the provided `select` function is of type list or function. WDYT?

**Comment by TkDodo** [comment]
I don't see an actual use-case here. what do you want to combine? the result of multiple queries? Please provide a code example or even a codesandbox reproduction that shows what you want to do / what you want to combine.

**Reply by dudulasry** [reply]
For example, I may want to retrieve a specific nested attribute from my `react-query` data, e.g user address. Let's say my server response looks like this:

```
{
  "user": {
    "email": "email@example.com"
    "first_name": "John",
    "last_name": "Smith",
    "address": {
      "street": "Sesame street 35"
    }
  }
}
```

We might want to write a selector that first fetches the user and then fetches the address. We don't want to type `data.user` all the time, so we can create a selector for that:

```
const userSelector = (data) => data.user;
const streetSelector = (user) => user.address.street;

const Component = () => {
  const { street } = useQuery('fetchUser', fetchUser, {
    select: combineSelectors([userSelector, streetSelector]);
  });
}
```

The good thing (IMO) about this is that if we have a similar logic elsewhere in the application (for example fetching list of users and get their street names), we can reuse this piece of logic.

**Reply by TkDodo** [reply]
But combine them into what? One object? That would require each selector to return an object, which it might not do. So we need at least a combine function, and now we’ve just described `reselect`, which you can use very well together wirth RQ:

```
const userSelector = (data) => data.user;
const streetSelector = (user) => user.address.street;

const userDetailsSelector = createSelector(
  userSelector,
  streetSelector,
  (user, street) => ({ …user, …street })
)

const Component = () => {
  const { street } = useQuery('fetchUser', fetchUser, {
    select: userDetailsSelector
  });
}
```
